aea/helpers/async_utils.py

```diff
@@ -63,37 +64,54 @@ def ensure_list(value: Any) -> List:
 class AsyncState:
     """Awaitable state."""
 
-    def __init__(self, initial_state: Any = None):
+    def __init__(
+        self, initial_state: Any = None, states_enum: Optional[Container[Any]] = None
+    ):
         """Init async state.
 
         :param initial_state: state to set on start.
+        :param states_enum: container of valid states if not provided state not checked on set.
         """
         self._state = initial_state
         self._watchers: Set[Future] = set()
-
-    @property
-    def state(self) -> Any:
-        """Return current state."""
-        return self.get()
-
-    @state.setter
-    def state(self, state: Any) -> None:
-        """Set state."""
-        self.set(state)
+        self._callbacks: List[Callable] = []
+        self._states_enum = states_enum
 
     def set(self, state: Any) -> None:
         """Set state."""
+        if self._states_enum is not None and state not in self._states_enum:
+            raise ValueError(
+                f"Unsupported state: {state}. Valid states are {self._states_enum}"
+            )
+
         if self._state == state:  # pragma: no cover
             return
+
         self._state_changed(state)
         self._state = state
 
+    def add_callback(self, callback_fn: Callable) -> None:
```

Comment:
The same applies here re `Callable` annotation

Reply:
done